### PR TITLE
CircularLinkedList.java

### PR DESCRIPTION
**Pull request by prakhar96**
Added CircularLinkedList data structure.

### CODE REVIEW

File: src/main/java/com/lists/CircularLinkedList.java [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+ */
+package com.lists;
+
+
+import java.util.Iterator;
+
+/*
+* LinkedList provides following API
+        * ----------------------------------------------
+        *         LinkedList ()                   // constructs a empty list. O(1)
```
Comment (ashish-chopra) [inline src/main/java/com/lists/CircularLinkedList.java]: constructor name is  CircularLinkedList()

File: src/main/java/com/lists/CircularLinkedList.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+            current = current.next;
+            return item;
+        }
+
+    }
+    class LinkedListNode{
+        Item item;
+        LinkedListNode next;
+
+        LinkedListNode(Item i){
```
Comment (ashish-chopra) [inline src/main/java/com/lists/CircularLinkedList.java]: Just like other functions in this class use "item" as variable instead of "i".

File: src/main/java/com/lists/CircularLinkedList.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        * boolean remove(int index)	   // remove the node and return true at the given index from the list. O(n)
+        * Iterator<Item> iterator()   	   // Iterate over the linkedList
+        *
+*/
+
+public class CircularLinkedList<Item> implements Iterable<Item>{
+    LinkedListNode head, tail;
+    int size;
+
+    /*
```
Comment (ashish-chopra) [inline src/main/java/com/lists/CircularLinkedList.java]: Good attempt for commenting, but for public functions use java style commenting /**  */. Because we have a way to extract java doc which is very useful when we just want to browse the API in whole.

**Comment by ashish-chopra** [comment]
Hi Prakhar, Thanks for proceeding on the suggestion. By the way, i have given some basic cosmetic comments. And also write a Junit for this data structure in tests folder. You can refer to other data structures test cases. Make sure you test all the corner cases and even if you can introduce anything new in testing, which others have not yet added. Make the changes and the test case, i will then merge the request. Thanks.

File: src/main/java/com/lists/CircularLinkedList.java (hunk 3 above)
Comment (prakhar96) [inline src/main/java/com/lists/CircularLinkedList.java]: Sure .... I'll just make the required changes and send a pull request :)

File: src/main/java/com/lists/CircularLinkedList.java [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        returns an iterator
+        to list
+     */
+
+    public Iterator<Item> iterator(){
+        return new CircularLinkedListIterator();
+    }
+
+    class CircularLinkedListIterator implements Iterator<Item> {
```
Comment (ashish-chopra) [inline src/main/java/com/lists/CircularLinkedList.java]: Also make all inner classes inside this class "private". We do not want to others to access these classes at all.

**Comment by ashish-chopra** [comment]
Hi Prakhar,

Why dont you submit the josephus problem's code written using CircularLinkedList? Make it a seperate class.
